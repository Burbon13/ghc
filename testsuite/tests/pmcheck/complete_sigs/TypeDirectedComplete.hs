{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE ViewPatterns #-}
{-# OPTIONS_GHC -Wall #-}

module TypeDirectedComplete where

data Proxy a = Proxy

class IsEmpty a where
  isEmpty :: a -> Bool

class IsCons a where
  type Elt a
  isCons :: a -> Maybe (Elt a, a)

pattern Empty :: IsEmpty a => a
pattern Empty <- (isEmpty -> True)

pattern Cons :: IsCons a => Elt a -> a -> a
pattern Cons x xs <- (isCons -> Just (x,xs))

instance IsEmpty (Proxy a) where
  isEmpty Proxy = True

instance IsEmpty [a] where
  isEmpty = null

instance IsCons [a] where
  type Elt [a] = a
  isCons [] = Nothing
  isCons (x:xs) = Just (x,xs)

{-# COMPLETE Empty :: Proxy a #-}
{-# COMPLETE Empty, Cons :: [a] #-}

foo :: Proxy a -> Int
foo Empty = 0

bar :: [a] -> Int
bar Empty = 0
bar (Cons _ _) = 1

baz :: [a] -> Int
baz Empty = 0
