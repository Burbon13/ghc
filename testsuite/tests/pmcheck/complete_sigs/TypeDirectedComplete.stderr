
TypeDirectedComplete.hs:45:1: warning: [-Wincomplete-patterns (in -Wextra)]
    Pattern match(es) are non-exhaustive
    In an equation for ‘baz’:
        Patterns of type ‘[a]’ not matched: Cons _ _
